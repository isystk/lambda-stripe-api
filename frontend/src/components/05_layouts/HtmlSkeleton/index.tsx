import React, { Children, FC, useState } from 'react'
import { connect } from '@/components/hoc'
import { ContainerProps, WithChildren } from 'types'
import useAppRoot from '@/stores/useAppRoot'
import Title, { TitleProps } from './Title'
import Head from 'next/head'
import { isReactElement } from '@/utils/general/object'
import { APP_NAME, APP_DESCRIPTION } from '@/constants'

/** HtmlSkeleton Props */
export type HtmlSkeletonProps = WithChildren & {
  showSideMenu?: boolean
}
/** Presenter Props */
export type PresenterProps = HtmlSkeletonProps & {
  title?: TitleProps['children']
  description: string
}

/** Presenter Component */
const HtmlSkeletonPresenter: FC<PresenterProps> = ({
  showSideMenu,
  main,
  title,
  description,
  children,
}) => (
  <>
    <Head>
      {/* タイトル */}
      <title>{title}</title>
      {/* favicon */}
      <link rel="icon" href="/favicon.ico" />
      {/* PCやモバイル（スマホ、タブレット）などのデバイスごとのコンテンツの表示領域 */}
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      {/* ブラウザテーマカラー */}
      <meta name="theme-color" content="#ffffff" />
      {/* サイト概要 */}
      <meta name="description" content={description} />
      {/* OGP 画像URL */}
      <meta property="og:image" content={'/ogp-image.png'} />
      {/* OGP タイトル */}
      <meta name="og:title" content={title} />
      {/* OGP サイト概要 */}
      <meta name="og:description" content={description} />
      {/* OGP Twitterカード */}
      <meta name="twitter:card" content="summary" />
      {/* apple ポータブル端末 アイコン */}
      <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
      {/* manifest.json */}
      <link rel="manifest" href="/manifest.json" />
      {/* Tailwind UI で採用されているカスタムフォント */}
      <link rel="stylesheet" href="https://rsms.me/inter/inter.css" />
    </Head>
    <Context.Provider value={main}>{children}</Context.Provider>
  </>
)

/** Container Component */
const HtmlSkeletonContainer: React.FC<
  ContainerProps<HtmlSkeletonProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  let title: TitleProps['children'] | undefined = undefined
  const main = useAppRoot()

  if (!main) return <></>

  children = Children.map(children, (child) =>
    isReactElement(child) && child.type === Title
      ? (title = `${child.props.children} | ${APP_NAME}`) && undefined
      : child
  )
  if (!title) title = APP_NAME

  const description = APP_DESCRIPTION
  return presenter({
    main,
    title,
    description,
    children,
    ...props,
  })
}

export const Context = React.createContext(null)

export default connect<HtmlSkeletonProps, PresenterProps>(
  'HtmlSkeleton',
  HtmlSkeletonPresenter,
  HtmlSkeletonContainer
)

// Sub Component
export type { TitleProps }
export { Title }
