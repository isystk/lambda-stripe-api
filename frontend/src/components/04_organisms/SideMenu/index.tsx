import React, { FC, useContext } from 'react'
import { ContainerProps, WithChildren } from 'types'
import * as styles from './styles'
import { connect } from '@/components/hoc'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import MainService from '@/services/main'
import { Url } from '@/constants/url'

/** SideMenuProps Props */
export type SideMenuProps = WithChildren & { setMenuOpen; isMenuOpen }
/** Presenter Props */
export type PresenterProps = SideMenuProps & {
  main
  menu
  setMenuOpen
  isMenuOpen
}

/** Presenter Component */
const SideMenuPresenter: FC<PresenterProps> = ({
  main,
  menu,
  setMenuOpen,
  isMenuOpen,
  ...props
}) => (
  <>
    <div
      className={`${
        styles.sideMenu
      } fixed top-0 right-0 z-30 bg-white h-screen w-64 ${
        isMenuOpen ? styles.menuOpen : styles.menuClose
      }`}
    >
      <div className="mt-24"></div>
      <ul className="list-none">
        <li className="py-4">
          <a
            href={Url.Top}
            className="block text-gray-700 font-bold whitespace-nowrap px-4"
          >
            ホーム
          </a>
        </li>
        <li className="py-4">
          <a
            href={Url.Payment}
            className="block text-gray-600 hover:text-gray-700 whitespace-nowrap px-4"
          >
            料金
          </a>
        </li>
        <li className="py-4">
          <a
            href="https://twitter.com/ise0615"
            target="_blank"
            rel="noreferrer"
            className="block text-gray-600 hover:text-gray-700 whitespace-nowrap px-4"
          >
            お問い合わせ
          </a>
        </li>
      </ul>
    </div>
    <div
      className={`${
        isMenuOpen ? 'block' : 'hidden'
      } z-20 bg-black opacity-50 fixed top-0 w-full h-screen`}
      onClick={() => setMenuOpen(!isMenuOpen)}
    ></div>
  </>
)

/** Container Component */
const SideMenuContainer: React.FC<
  ContainerProps<SideMenuProps, PresenterProps>
> = ({ presenter, children, setMenuOpen, isMenuOpen, ...props }) => {
  const main = useContext<MainService | null>(Context)
  if (!main) return <></>

  return presenter({
    children,
    main,
    setMenuOpen,
    isMenuOpen,
    ...props,
  })
}

export default connect<SideMenuProps, PresenterProps>(
  'SideMenu',
  SideMenuPresenter,
  SideMenuContainer
)
