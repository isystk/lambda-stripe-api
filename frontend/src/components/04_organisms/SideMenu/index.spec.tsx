import React from 'react'
import renderer from 'react-test-renderer'
import SideMenu from './index'

import { Context } from '@/components/05_layouts/HtmlSkeleton'
import MainService from '@/services/main'

describe('SideMenu', () => {
  it('Match Snapshot', () => {
    const main = new MainService(() => ({}))

    const component = renderer.create(
      <Context.Provider value={main}>
        <SideMenu
          isMenuOpen={false}
          setMenuOpen={(isOpen) => console.log(isOpen)}
        />
      </Context.Provider>
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
