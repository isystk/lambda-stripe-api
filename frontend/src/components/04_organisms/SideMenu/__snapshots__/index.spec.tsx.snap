// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu Match Snapshot 1`] = `
[
  <div
    className="css-0 fixed top-0 right-0 z-30 bg-white h-screen w-64 css-hc9ocb"
  >
    <div
      className="mt-24"
    />
    <ul
      className="list-none"
    >
      <li
        className="py-4"
      >
        <a
          className="block text-gray-700 font-bold whitespace-nowrap px-4"
          href="/"
        >
          ホーム
        </a>
      </li>
      <li
        className="py-4"
      >
        <a
          className="block text-gray-600 hover:text-gray-700 whitespace-nowrap px-4"
          href="/product/prod_NpvV9ohJIlgElI/payment"
        >
          料金
        </a>
      </li>
      <li
        className="py-4"
      >
        <a
          className="block text-gray-600 hover:text-gray-700 whitespace-nowrap px-4"
          href="https://twitter.com/ise0615"
          rel="noreferrer"
          target="_blank"
        >
          お問い合わせ
        </a>
      </li>
    </ul>
  </div>,
  <div
    className="hidden z-20 bg-black opacity-50 fixed top-0 w-full h-screen"
    onClick={[Function]}
  />,
]
`;
